### PR TITLE
#1-4Task. Azizxon_uz

### PR DESCRIPTION
**Pull request by Azizxon**
https://travis-ci.com/github/Azizxon/js-assignments/builds/173730142

### CODE REVIEW

File: task/01-strings-tasks.js [hunk 1]
```diff
@@ -22,7 +22,7 @@
  *   '',  'bb'  => 'bb'
  */
 function concatenateStrings(value1, value2) {
-    throw new Error('Not implemented');
```
Comment (vgandzyuck) [inline task/01-strings-tasks.js:25]: OPTIONAL: also u can use template strings
https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/template_strings

File: task/01-strings-tasks.js [hunk 2]
```diff
@@ -201,7 +202,18 @@ function extractEmails(str) {
  *
  */
 function getRectangleString(width, height) {
-    throw new Error('Not implemented');
+    var top = '┌'.concat('─'.repeat(width - 2)).concat('┐');
+    var between = '│'.concat(' '.repeat(width - 2)).concat('│');
+    var bottom = "└".concat('─'.repeat(width - 2)).concat('┘');
+
+    var rect = top.concat('\n');
+    for (let index = 0; index < height - 2; index++) {
```
Comment (vgandzyuck) [inline task/01-strings-tasks.js:210]: u can avoid of using loop here

File: task/01-strings-tasks.js [hunk 3]
```diff
@@ -201,7 +202,18 @@ function extractEmails(str) {
  *
  */
 function getRectangleString(width, height) {
-    throw new Error('Not implemented');
+    var top = '┌'.concat('─'.repeat(width - 2)).concat('┐');
+    var between = '│'.concat(' '.repeat(width - 2)).concat('│');
+    var bottom = "└".concat('─'.repeat(width - 2)).concat('┘');
+
+    var rect = top.concat('\n');
```
Comment (vgandzyuck) [inline task/01-strings-tasks.js:209]: why u did't concated new line character at line 205?

File: task/01-strings-tasks.js [hunk 4]
```diff
@@ -201,7 +202,18 @@ function extractEmails(str) {
  *
  */
 function getRectangleString(width, height) {
-    throw new Error('Not implemented');
+    var top = '┌'.concat('─'.repeat(width - 2)).concat('┐');
```
Comment (vgandzyuck) [inline task/01-strings-tasks.js:205]: Please, try to use only this list of statements and methods

1. const
2. concat
3. repeat
4. return :)

File: task/01-strings-tasks.js [hunk 5]
```diff
@@ -221,7 +233,9 @@ function getRectangleString(width, height) {
  *
  */
 function encodeToRot13(str) {
-    throw new Error('Not implemented');
+    return (str + '').replace(/[a-zA-Z]/gi, function (s) {
```
Comment (vgandzyuck) [inline task/01-strings-tasks.js:236]: please use [arrow function](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Functions/Arrow_functions) instead

File: task/01-strings-tasks.js [hunk 6]
```diff
@@ -238,7 +252,7 @@ function encodeToRot13(str) {
  *   isString(new String('test')) => true
  */
 function isString(value) {
-    throw new Error('Not implemented');
+    return (typeof value === 'string' || value instanceof String) ? true : false;
```
Comment (vgandzyuck) [inline task/01-strings-tasks.js:255]: There are is no need need to use ternary operator here, in this case [logical operator OR](https://javascript.info/logical-operators) will return boolean value:)

File: task/02-numbers-tasks.js [hunk 7]
```diff
@@ -54,7 +54,7 @@ function getCicleCircumference(radius) {
  *  -3, 3  => 0
  */
 function getAverage(value1, value2) {
-    throw new Error('Not implemented');
+    return value1 / 2 + value2 / 2;
```
Comment (vgandzyuck) [inline task/02-numbers-tasks.js:57]: division is a heavy operation, try to reduce total count calls of it

File: task/02-numbers-tasks.js [hunk 8]
```diff
@@ -202,7 +208,10 @@ function roundToPowerOfTen(num, pow) {
  *   17 => true
  */
 function isPrime(n) {
-    throw new Error('Not implemented');
+    for (let i = 2, s = Math.sqrt(n); i <= s; i++) {
+        if (n % i === 0) return false;
```
Comment (vgandzyuck) [inline task/02-numbers-tasks.js:212]: Please, always use curly braces with "if" statement

File: task/03-date-tasks.js [hunk 9]
```diff
@@ -56,7 +56,8 @@ function parseDataFromIso8601(value) {
  *    Date(2015,1,1)    => false
  */
 function isLeapYear(date) {
-   throw new Error('Not implemented');
+   var year = date.getFullYear();
+   return (year % 400 == 0 || year % 100 != 0) && (year % 4 == 0) ? true : false;
```
Comment (vgandzyuck) [inline task/03-date-tasks.js:60]: As I mentioned before, no need to use ternary here)

File: test/03-date-tests.js [hunk 10]
```diff
@@ -86,7 +86,7 @@ describe('03-date-tasks', function() {
             }, {
                 startDate: new Date(2000,1,1,10,0,0),
                 endDate:   new Date(2000,1,2,15,20,10,453),
-                expected:  '29:20:10.453'
```
Comment (vgandzyuck) [inline test/03-date-tests.js:89]: Please, revert this changes

File: task/01-strings-tasks.js [hunk 11]
```diff
@@ -69,7 +69,8 @@ function getStringFromTemplate(firstName, lastName) {
  *   'Hello, Chuck Norris!' => 'Chuck Norris'
  */
 function extractNameFromTemplate(value) {
-    throw new Error('Not implemented');
+    var temp = value.substring(7);
```
[the remark this answers is not in the document]
Comment (Azizxon) [reply task/01-strings-tasks.js]: ok, replace any possible where

File: task/01-strings-tasks.js [hunk 12]
```diff
@@ -69,7 +69,8 @@ function getStringFromTemplate(firstName, lastName) {
...
+    return temp.substring(0, temp.length - 1);
```
[the remark this answers is not in the document]
Comment (Azizxon) [reply task/01-strings-tasks.js]: yeap, fixed

File: test/03-date-tests.js (hunk 10 above)
Comment (Azizxon) [reply test/03-date-tests.js]: fixed, I recently saw correct date difference

File: task/01-strings-tasks.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (IlyaMokin) [reply task/01-strings-tasks.js]: LGTM, but one more small note, substring can also work from the end of string.
Example: `temp.substring(0, -1)` - give you the same result

**Comment by IlyaMokin** [comment]
Acepted!

**Comment by IlyaMokin** [comment]
Not a big deal, but array-tests file has cahnges still.

Also the PR has redundant changes in styling, spaces, comments....
Please follow simple rules:
 - do not add any non logic changes to code if review is required
 - if the changes in styling are important - you can do it as separate PR with styling only changes to avoid noise in the main PR

**Comment by Azizxon** [comment]
done

**Comment by IlyaMokin** [comment]
Looks cool! Accepted!
